### PR TITLE
Parse snippet of repodata (RhBug:1804308) and remove expat

### PR DESCRIPTION
**Pull request by kontura**
Add functions cr_xml_parse_(primary/filelists/other)_snippet and their python counterparts to the API.

These functions allow parsing of repodata passed as a string (instead of a path to file) which doesn't contain a root xml element (eg. `<filelists>`).

There are also tests for the new functionality.

https://bugzilla.redhat.com/show_bug.cgi?id=1804308

### CODE REVIEW

File: src/xml_parser.h [hunk 1]
```diff
@@ -171,6 +223,31 @@ int cr_xml_parse_other(const char *path,
...
+ * @param pkgcb          Package callback. (Called when complete package
+ *                       xml chunk is parsed.). Could be NULL if newpkgcb is
+ *                       not NULL.
+ * @param pkgcb_data     User data for the pkgcb.
+ * @param warningcb      Callback for warning messages.
+ * @param warningcb_data User data for the warningcb.
+ * @param err            GError **
+ * @return               cr_Error code.
+ */
+int cr_xml_parse_other_snippet(const char *path,
```
Comment (dralley) [inline src/xml_parser.h]: Same here

File: src/xml_parser.h [hunk 2]
```diff
@@ -123,6 +123,33 @@ int cr_xml_parse_primary(const char *path,
...
+ *                       xml chunk is parsed.). Could be NULL if newpkgcb is
+ *                       not NULL.
+ * @param pkgcb_data     User data for the pkgcb.
+ * @param warningcb      Callback for warning messages.
+ * @param warningcb_data User data for the warningcb.
+ * @param do_files       0 - Ignore file tags in primary.xml.
+ * @param err            GError **
+ * @return               cr_Error code.
+ */
+int cr_xml_parse_primary_snippet(const char *path,
```
Comment (dralley) [inline src/xml_parser.h]: Signature doesn't match the documentation, this should probably be ```xml_string```

File: src/xml_parser.h (hunk 1 above)
Comment (kontura) [reply src/xml_parser.h]: Thanks, I have fixed these.

**Comment by kontura** [comment]
Since I was working with the xml parser I took the opportunity and switched the expat parser for libxml2.
Resolves: https://github.com/rpm-software-management/createrepo_c/issues/150
I did some tests with big files and the speeds were comparable. It doesn't seem particularly faster but also not slower.

**Comment by Conan-Kudo** [comment]
@kontura I don't know if I'm reading this correctly, but it _seems_ like the unit tests for the Python function for parsing repodata snippets is _before_ the implementation of the function itself? Does it make sense to have them split in that order? Also, does it make sense to split the implementation of the code and tests into separate commits this way, rather than having the function and its tests in the same commit?

**Comment by Conan-Kudo** [comment]
Nope, nevermind, GitHub's rendering of this is just dumb.

**Comment by dralley** [comment]
So I'm not sure if one of the later changes broke something that had previously been working, but it appears that all of the snippet tests are failing.  

Screenshot from 2020-03-11 12-51-40

Screenshot from 2020-03-11 12-57-03

It appears to be trying to fetch the XML string contents though it were a path.  It's not just an issue with the test, I can also confirm this with my own testing.

**Comment by dralley** [comment]
disregard my last comment

**Pull request merged**